File: src/lib/scrapers/bookmyshow.ts
import { chromium } from "playwright";

export type ScrapedEvent = {
  title: string;
  eventName: string;
  platform: "BookMyShow" | "District" | "SortMyScene" | "Other";
  eventType: string;
  city: string;
  venue: string;
  eventLink: string;
  startDate?: string;
};

const CITY_SLUGS: Record<string, string> = {
  bangalore: "bangalore",
  bengaluru: "bangalore",
  mumbai: "mumbai",
  delhi: "ncr",
  hyderabad: "hyderabad",
  chennai: "chennai",
  pune: "pune",
  kolkata: "kolkata",
  ahmedabad: "ahmedabad",
  goa: "goa",
};

export async function scrapeBookMyShow(city: string): Promise<ScrapedEvent[]> {
  const citySlug = CITY_SLUGS[city.toLowerCase()] ?? city.toLowerCase();
  const url = `https://in.bookmyshow.com/explore/events-${citySlug}`;

  const browser = await chromium.launch({
    headless: true,
    args: [
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-blink-features=AutomationControlled",
      "--disable-web-security",
    ],
  });

  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
    viewport: { width: 1280, height: 900 },
    extraHTTPHeaders: {
      "Accept-Language": "en-IN,en;q=0.9",
    },
  });

  const page = await context.newPage();
  const events: ScrapedEvent[] = [];

  try {
    await page.goto(url, { waitUntil: "networkidle", timeout: 45000 });

    // Give JS extra time to render
    await page.waitForTimeout(3000);

    // Scroll to trigger lazy loading
    for (let i = 0; i < 3; i++) {
      await page.evaluate(() => window.scrollBy(0, window.innerHeight));
      await page.waitForTimeout(1000);
    }

    // ✅ Dump ALL links and filter by BMS event URL pattern
    const allLinks = await page.evaluate(() => {
      return Array.from(document.querySelectorAll("a[href]")).map((a) => {
        const el = a as HTMLAnchorElement;
        const href = el.href;

        // Walk up the DOM to find text content near this link
        const text = el.innerText?.trim() ||
          el.querySelector("*")?.textContent?.trim() || "";

        // Try to find an image alt as title fallback
        const imgAlt = el.querySelector("img")?.alt?.trim() || "";

        return {
          href,
          text: text || imgAlt,
        };
      });
    });

    // Filter to only event links
    const eventLinks = allLinks.filter(
      (l) =>
        l.href.includes("bookmyshow.com") &&
        (l.href.includes("/events/") ||
          l.href.includes("/activities/") ||
          l.href.includes("/show/")) &&
        l.text.length > 2 &&
        !l.href.includes("explore") &&
        !l.href.includes("undefined")
    );

    const seen = new Set<string>();
    for (const link of eventLinks) {
      if (seen.has(link.href)) continue;
      seen.add(link.href);

      // Extract event name from URL slug as fallback
      const urlSlug = link.href
        .split("/")
        .pop()
        ?.split("?")[0]
        ?.replace(/-/g, " ") ?? "";

      events.push({
        title: link.text || urlSlug,
        eventName: link.text || urlSlug,
        platform: "BookMyShow",
        eventType: "Event",
        city,
        venue: "—",
        eventLink: link.href,
      });
    }

    console.log(`BMS: found ${eventLinks.length} event links on page`);
  } catch (err) {
    console.error("BMS scraper error:", err);
  } finally {
    await browser.close();
  }

  return events;
}

File: scraper/scrapers/bookmyshow.py
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from webdriver_manager.chrome import ChromeDriverManager
from bs4 import BeautifulSoup
import time
import re
import os


BMS_CITY_SLUGS = {
    "bangalore": "bangalore",
    "bengaluru": "bangalore",
    "mumbai": "mumbai",
    "delhi": "ncr",
    "hyderabad": "hyderabad",
    "chennai": "chennai",
    "pune": "pune",
    "kolkata": "kolkata",
    "ahmedabad": "ahmedabad",
    "goa": "goa",
}


def _get_driver():
    options = Options()

    # keep headless off for debugging first
    # options.add_argument("--headless=new")

    options.add_argument("--no-sandbox")
    options.add_argument("--disable-dev-shm-usage")
    options.add_argument("--disable-gpu")
    options.add_argument("--window-size=1400,2200")
    options.add_argument("--disable-blink-features=AutomationControlled")
    options.add_experimental_option("excludeSwitches", ["enable-automation"])
    options.add_experimental_option("useAutomationExtension", False)
    options.add_argument(
        "--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36"
    )

    service = Service(ChromeDriverManager().install())
    driver = webdriver.Chrome(service=service, options=options)
    driver.set_page_load_timeout(60)
    return driver


def _clean_text(text: str) -> str:
    if not text:
        return ""
    text = text.replace("\n", " ").replace("\r", " ")
    text = re.sub(r"\s+", " ", text).strip()
    text = re.sub(r"\s*book tickets?.*$", "", text, flags=re.IGNORECASE).strip()
    return text


def _valid_bms_link(href: str) -> bool:
    if not href:
        return False
    if href.startswith("/"):
        href = f"https://in.bookmyshow.com{href}"
    if "bookmyshow.com" not in href:
        return False
    if "/explore/" in href:
        return False
    if not any(x in href for x in ["/events/", "/activities/", "/plays/", "/sports/", "/comedy-shows/", "/music-shows/"]):
        return False
    return True


def _parse_bms_html(html: str, city: str) -> list[dict]:
    soup = BeautifulSoup(html, "html.parser")
    events = []
    seen = set()

    for a in soup.find_all("a", href=True):
        href = a["href"].strip()

        if href.startswith("/"):
            href = f"https://in.bookmyshow.com{href}"

        if not _valid_bms_link(href):
            continue

        if href in seen:
            continue
        seen.add(href)

        title = _clean_text(a.get_text(" ", strip=True))
        if len(title) < 4:
            img = a.find("img")
            if img:
                title = _clean_text(img.get("alt", ""))

        if len(title) < 4:
            slug = href.rstrip("/").split("/")[-1].split("?")[0]
            title = re.sub(r"[-_]+", " ", slug).strip().title()

        if len(title) < 4:
            continue

        events.append({
            "title": title[:140],
            "eventName": title[:140],
            "platform": "BookMyShow",
            "eventType": "Event",
            "city": city,
            "venue": "—",
            "eventLink": href,
            "startDate": None,
        })

    return events


def scrape_bookmyshow(city: str) -> list[dict]:
    print(f"[BMS] Scraping for city: {city}")
    slug = BMS_CITY_SLUGS.get(city.lower(), city.lower())
    url = f"https://in.bookmyshow.com/explore/events-{slug}"
    print(f"[BMS] URL: {url}")

    os.makedirs("debug_html", exist_ok=True)

    driver = None
    try:
        driver = _get_driver()
        driver.get(url)

        WebDriverWait(driver, 30).until(
            EC.presence_of_element_located((By.TAG_NAME, "body"))
        )

        time.sleep(6)

        for _ in range(8):
            driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
            time.sleep(2)

        print("[BMS] Current URL:", driver.current_url)
        print("[BMS] Page title:", driver.title)

        html = driver.page_source
        with open("debug_html/bms.html", "w", encoding="utf-8") as f:
            f.write(html)

        events = _parse_bms_html(html, city)
        print(f"[BMS] Found {len(events)} events")
        return events

    except Exception as e:
        print(f"[BMS] Failed: {e}")
        raise
    finally:
        if driver:
            driver.quit()